test/hotspot/jtreg/gc/TestJNICriticalStressTest.java

```diff
@@ -0,0 +1,236 @@
+        private int count;
+
+        private void doStep() {
+            byte[] inputArray = new byte[BYTE_ARRAY_LENGTH];
+            for (int i = 0; i < inputArray.length; i += 1) {
+                inputArray[i] = (byte) (count + i);
+            }
+
+            Deflater deflater = new Deflater();
+            deflater.setInput(inputArray);
+            deflater.finish();
+
+            byte[] outputArray = new byte[2 * inputArray.length];
+            deflater.deflate(outputArray);
+
+            count += 1;
+        }
+
+        public void run() {
+            while (true) {
+                doStep();
+            }
+        }
+    }
+
+    static private class SystemGCWorker implements Runnable {
+        public void run() {
+            while (true) {
+                try {
+                    Thread.sleep(SYSTEM_GC_PERIOD_MS);
+                } catch (InterruptedException e) {
+                }
+                System.out.println("SYSTEM_GC BEFORE");
+                wb.fullGC();
+                fullGcCounts += 1;
+                System.out.println("SYSTEM_GC AFTER");
+            }
+        }
+    }
+
+    static public Map<Integer,String> largeMap;
+
+    public static void main(String... args) throws Exception {
+        if (args.length < 4) {
+            println("usage: JNICriticalStressTest3 <duration sec> <alloc threads> <jni critical threads>");
+            exit(-1);
+        }
+
+        long durationSec = Long.parseLong(args[0]);
+        int allocThreadNum = Integer.parseInt(args[1]);
+        int jniCriticalThreadNum = Integer.parseInt(args[2]);
+
+        StringBuilder OldGCName = new StringBuilder();
+        switch (args[3]) {
+            case "G1":
+                OldGCName.append("G1 Old Generation");
+                break;
+            case "Parallel":
+                OldGCName.append("PS MarkSweep");
+                break;
+            case "Serial":
+                OldGCName.append("MarkSweepCompact");
+                break;
+            default:
+                throw new RuntimeException("Unsupported GC selected");
+        }
+
+        List<GarbageCollectorMXBean> collectors = ManagementFactory.getGarbageCollectorMXBeans();
+
+        GarbageCollectorMXBean collector = null;
+
+        for (int i = 0; i < collectors.size(); i++) {
+            println(collectors.get(i).getName());
+            if (collectors.get(i).getName().contains(OldGCName.toString())) {
+                collector = collectors.get(i);
+                break;
+            }
+        }
+
+        if (collector == null) {
+            throw new RuntimeException(OldGCName.toString() + " not found");
+        }
+
+        println("Running for " + durationSec + " secs");
+
+        largeMap = populateMap(LARGE_MAP_SIZE);
+
+        println("Starting " + allocThreadNum + " allocating threads");
+        for (int i = 0; i < allocThreadNum; i += 1) {
+            new Thread(new AllocatingWorker()).start();
```

Comment:
What's the point of having an alloc thread? I'd expect whiteboxapi to be enough to trigger gc cycles regardless of heap state.

Reply:
https://bugs.openjdk.org/browse/JDK-8057573

Reproducer taken from here. Added minor changes to fit JTREG.

The alloc thread test makes for more robust testing, but not required.

Reply:
I don't get why having an alloc thread would improve/degrade the robustness of this test -- jni-call-thread and systemgc-thread should be enough.

Reply:
Please add a comment that this is intentional to increase the variety of the GC types that can happen here.